[NT] Make NestedTensor register as having symbolic sizes/strides

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124687

Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. It passes in the actual sizes when we call `_make_wrapper_subclass` - this is the change that makes the subclass register as `has_symbolic_sizes_strides() == True`
2. It adds a field to `_make_wrapper_subclass` where an explicit `numel` can be provided. This allows us to skip the numel computation for the storage, which previously fails due to arithmetic on NestedInts.
3. Implements `aten::numel` for NJT - this is separate from the overridden numel in `make_wrapper_subclass` for now. Note also that this means that we leave `dispatch_sizes_strides_policy="sizes"`, so that we call into the custom `numel` implementation (as well as `sizes` and `strides`), because `numel` cannot currently be computed from `sizes` for NJT.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

Differential Revision: [D57225736](https://our.internmc.facebook.com/intern/diff/D57225736)

davidberard98 committed:
[WIP] Make NestedTensor register as having symbolic sizes/strides

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] Make NestedTensor register as having symbolic sizes/strides"




[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] Make NestedTensor register as having symbolic sizes/strides"




[ghstack-poisoned]

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. Add python bindings for `torch.Tensor._set_sizes_and_strides()`. **Note: Do we want to make this available from python? Should we name it "unsafe" or something?**
2. Call `self._set_sizes_and_strides()` on NT during `__init__`, which sets the size to be symbolic. **Note: is this fine - would it be better to put the size/stride calculations in `__new__` when we initially construct the wrapper subclass?**
3. In `__new__` when we call `_make_wrapper_subclass`, we _still_ set `dispatch_sizes_strides_policy="sizes"`. This forces sizes, strides, and numel to call into python; in particular, we need this for `numel`, where we want the custom implementation that looks at the values tensor's numel. **Note: we only actually need to call a custom numel.  We could probably clean this up further and change the sizes/strides policy stuff, which seems like it only exists for NT?**
4. Implement `numel` in ops.py (`sym_numel` is already implemented, but the numel value is not symbolic so we should also be able to implement `numel`). This makes `nt.numel()` return a valid value.

[ghstack-poisoned]

soulitzer commented:
> Add python bindings for torch.Tensor._set_sizes_and_strides(). Note: Do we want to make this available from python? Should we name it "unsafe" or something?
Call self._set_sizes_and_strides() on NT during __init__, which sets the size to be symbolic. Note: is this fine - would it be better to put the size/stride calculations in __new__ when we initially construct the wrapper subclass?

hmm yeah I think it is preferred to do more things in new/_make_wrapper_subclass, and that should avoid having to call _set_sizes_and_strides manually ourselves.

> in particular, we need this for numel, where we want the custom implementation that looks at the values tensor's numel. 

To fix this, we'd need to make numel as a field stored somewhere rather than go through the default logic of multiplying all the sizes.

davidberard98 committed:
Try pushing the shape/stride logic directly into make_wrapper_subclass on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. Add python bindings for `torch.Tensor._set_sizes_and_strides()`. **Note: Do we want to make this available from python? Should we name it "unsafe" or something?**
2. Call `self._set_sizes_and_strides()` on NT during `__init__`, which sets the size to be symbolic. **Note: is this fine - would it be better to put the size/stride calculations in `__new__` when we initially construct the wrapper subclass?**
3. In `__new__` when we call `_make_wrapper_subclass`, we _still_ set `dispatch_sizes_strides_policy="sizes"`. This forces sizes, strides, and numel to call into python; in particular, we need this for `numel`, where we want the custom implementation that looks at the values tensor's numel. **Note: we only actually need to call a custom numel.  We could probably clean this up further and change the sizes/strides policy stuff, which seems like it only exists for NT?**
4. Implement `numel` in ops.py (`sym_numel` is already implemented, but the numel value is not symbolic so we should also be able to implement `numel`). This makes `nt.numel()` return a valid value.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

albanD commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -76,31 +98,16 @@ def __new__(
             False,
             True,  # dispatch_layout
             ks,
+            _allocate_storage=False,  # don't try to allocate storage based on non-zero size
```
Don't you have a valid size/stride/offset values here? Why don't you want to allocate these? This is a fake storage for show and alias resolution, it doesn't actually use any memory ever.

davidberard98 replied:
The storage size calculation fails with NestedInt sizes.

I didn't realize it was a fake storage though. Is it okay to have the zero-size storage used in this PR to avoid needing to deal with the NestedInt size or do you think there's other issues that will arise from that?

albanD replied:
I is better to have a storage, even if it is 0 size because it allows code that check for view-ness and stuff like that through storage to work.

davidberard98 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -631,6 +631,29 @@ static PyObject* THPVariable_view_func_unsafe(
   return view_func_impl(self_, args, kwargs, /*check_has_same_meta=*/false);
 }
 
+static PyObject* THPVariable_set_sizes_and_strides(
```
I just changed the PR to not use this. I'm planning to remove it if we're okay with this approach, but didn't want to delete it while the PR is WIP

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. Add python bindings for `torch.Tensor._set_sizes_and_strides()`. **Note: Do we want to make this available from python? Should we name it "unsafe" or something?**
2. Call `self._set_sizes_and_strides()` on NT during `__init__`, which sets the size to be symbolic. **Note: is this fine - would it be better to put the size/stride calculations in `__new__` when we initially construct the wrapper subclass?**
3. In `__new__` when we call `_make_wrapper_subclass`, we _still_ set `dispatch_sizes_strides_policy="sizes"`. This forces sizes, strides, and numel to call into python; in particular, we need this for `numel`, where we want the custom implementation that looks at the values tensor's numel. **Note: we only actually need to call a custom numel.  We could probably clean this up further and change the sizes/strides policy stuff, which seems like it only exists for NT?**
4. Implement `numel` in ops.py (`sym_numel` is already implemented, but the numel value is not symbolic so we should also be able to implement `numel`). This makes `nt.numel()` return a valid value.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. Add python bindings for `torch.Tensor._set_sizes_and_strides()`. **Note: Do we want to make this available from python? Should we name it "unsafe" or something?**
2. Call `self._set_sizes_and_strides()` on NT during `__init__`, which sets the size to be symbolic. **Note: is this fine - would it be better to put the size/stride calculations in `__new__` when we initially construct the wrapper subclass?**
3. In `__new__` when we call `_make_wrapper_subclass`, we _still_ set `dispatch_sizes_strides_policy="sizes"`. This forces sizes, strides, and numel to call into python; in particular, we need this for `numel`, where we want the custom implementation that looks at the values tensor's numel. **Note: we only actually need to call a custom numel.  We could probably clean this up further and change the sizes/strides policy stuff, which seems like it only exists for NT?**
4. Implement `numel` in ops.py (`sym_numel` is already implemented, but the numel value is not symbolic so we should also be able to implement `numel`). This makes `nt.numel()` return a valid value.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. Add python bindings for `torch.Tensor._set_sizes_and_strides()`. **Note: Do we want to make this available from python? Should we name it "unsafe" or something?**
2. Call `self._set_sizes_and_strides()` on NT during `__init__`, which sets the size to be symbolic. **Note: is this fine - would it be better to put the size/stride calculations in `__new__` when we initially construct the wrapper subclass?**
3. In `__new__` when we call `_make_wrapper_subclass`, we _still_ set `dispatch_sizes_strides_policy="sizes"`. This forces sizes, strides, and numel to call into python; in particular, we need this for `numel`, where we want the custom implementation that looks at the values tensor's numel. **Note: we only actually need to call a custom numel.  We could probably clean this up further and change the sizes/strides policy stuff, which seems like it only exists for NT?**
4. Implement `numel` in ops.py (`sym_numel` is already implemented, but the numel value is not symbolic so we should also be able to implement `numel`). This makes `nt.numel()` return a valid value.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. Add python bindings for `torch.Tensor._set_sizes_and_strides()`. **Note: Do we want to make this available from python? Should we name it "unsafe" or something?**
2. Call `self._set_sizes_and_strides()` on NT during `__init__`, which sets the size to be symbolic. **Note: is this fine - would it be better to put the size/stride calculations in `__new__` when we initially construct the wrapper subclass?**
3. In `__new__` when we call `_make_wrapper_subclass`, we _still_ set `dispatch_sizes_strides_policy="sizes"`. This forces sizes, strides, and numel to call into python; in particular, we need this for `numel`, where we want the custom implementation that looks at the values tensor's numel. **Note: we only actually need to call a custom numel.  We could probably clean this up further and change the sizes/strides policy stuff, which seems like it only exists for NT?**
4. Implement `numel` in ops.py (`sym_numel` is already implemented, but the numel value is not symbolic so we should also be able to implement `numel`). This makes `nt.numel()` return a valid value.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -76,31 +98,17 @@ def __new__(
             True,  # dispatch_layout
             ks,
+            # don't try to calculate storage based on non-zero size
+            _override_storage_size=values.numel() * values.dtype.itemsize,
```
Not sure where this matters tbh, but what about `values.untyped_storage().size()`?

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. Add python bindings for `torch.Tensor._set_sizes_and_strides()`. **Note: Do we want to make this available from python? Should we name it "unsafe" or something?**
2. Call `self._set_sizes_and_strides()` on NT during `__init__`, which sets the size to be symbolic. **Note: is this fine - would it be better to put the size/stride calculations in `__new__` when we initially construct the wrapper subclass?**
3. In `__new__` when we call `_make_wrapper_subclass`, we _still_ set `dispatch_sizes_strides_policy="sizes"`. This forces sizes, strides, and numel to call into python; in particular, we need this for `numel`, where we want the custom implementation that looks at the values tensor's numel. **Note: we only actually need to call a custom numel.  We could probably clean this up further and change the sizes/strides policy stuff, which seems like it only exists for NT?**
4. Implement `numel` in ops.py (`sym_numel` is already implemented, but the numel value is not symbolic so we should also be able to implement `numel`). This makes `nt.numel()` return a valid value.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. Add python bindings for `torch.Tensor._set_sizes_and_strides()`. **Note: Do we want to make this available from python? Should we name it "unsafe" or something?**
2. Call `self._set_sizes_and_strides()` on NT during `__init__`, which sets the size to be symbolic. **Note: is this fine - would it be better to put the size/stride calculations in `__new__` when we initially construct the wrapper subclass?**
3. In `__new__` when we call `_make_wrapper_subclass`, we _still_ set `dispatch_sizes_strides_policy="sizes"`. This forces sizes, strides, and numel to call into python; in particular, we need this for `numel`, where we want the custom implementation that looks at the values tensor's numel. **Note: we only actually need to call a custom numel.  We could probably clean this up further and change the sizes/strides policy stuff, which seems like it only exists for NT?**
4. Implement `numel` in ops.py (`sym_numel` is already implemented, but the numel value is not symbolic so we should also be able to implement `numel`). This makes `nt.numel()` return a valid value.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. Add python bindings for `torch.Tensor._set_sizes_and_strides()`. **Note: Do we want to make this available from python? Should we name it "unsafe" or something?**
2. Call `self._set_sizes_and_strides()` on NT during `__init__`, which sets the size to be symbolic. **Note: is this fine - would it be better to put the size/stride calculations in `__new__` when we initially construct the wrapper subclass?**
3. In `__new__` when we call `_make_wrapper_subclass`, we _still_ set `dispatch_sizes_strides_policy="sizes"`. This forces sizes, strides, and numel to call into python; in particular, we need this for `numel`, where we want the custom implementation that looks at the values tensor's numel. **Note: we only actually need to call a custom numel.  We could probably clean this up further and change the sizes/strides policy stuff, which seems like it only exists for NT?**
4. Implement `numel` in ops.py (`sym_numel` is already implemented, but the numel value is not symbolic so we should also be able to implement `numel`). This makes `nt.numel()` return a valid value.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

davidberard98 commented:
@albanD @soulitzer could I get another review? I've updated this to 
(a) not have a separate API for set_sizes_strides
(b) `_make_wrapper_subclass` now takes another optional argument for "override_numel", which if provided, overrides the storage size - based on @soulitzer's suggestion. Alternatively, I could make this just take `_override_storage_size`; is there a preference?

soulitzer commented:
> Alternatively, I could make this just take _override_storage_size; is there a preference?

hmm numel != storage_size, so technically override_storage_size is more accurate

davidberard98 commented:
@soulitzer on the other hand - (a) you point out that override_numel may be useful in the future, and (b) it sounds like we don't actually care about the storage since it is fake? do you think it's fine as it is or should I switch it back to override_storage_size?

davidberard98 commented:
ok, yeah I guess we might as well make it `override_storage_size` since that's what we're actually trying to do here - we can add numel and/or remove storage_size in the future if we decide we want to do that

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. It passes in the actual sizes when we call `_make_wrapper_subclass` - this is the change that makes the subclass register as `has_symbolic_sizes_strides() == True`
2. It adds a field to `_make_wrapper_subclass` where an explicit `numel` can be provided. This allows us to skip the numel computation for the storage, which previously fails due to arithmetic on NestedInts.
3. Implements `aten::numel` for NJT - this is separate from the overridden numel in `make_wrapper_subclass` for now. Note also that this means that we leave `dispatch_sizes_strides_policy="sizes"`, so that we call into the custom `numel` implementation (as well as `sizes` and `strides`), because `numel` cannot currently be computed from `sizes` for NJT.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot rebase

davidberard98 committed:
Update

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -751,9 +751,9 @@ static PyObject* THPVariable_make_wrapper_subclass(
       "Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False, "
       "c10::string_view? dispatch_sizes_strides_policy=None, bool dispatch_device=False, bool dispatch_layout=False, "
-      "DispatchKeySet _extra_dispatch_keys=None)",
+      "DispatchKeySet _extra_dispatch_keys=None, SymInt? _override_storage_size=None)",
```
```suggestion
      "DispatchKeySet _extra_dispatch_keys=None, SymInt? storage_size=None)",
```

- private arguments is not a thing, everything is visible to the user in the signature.
- We need to update the doc to explain what this argument is doing: set the storage size, if not provided, we use the smallest storage size possible based on the other arguments

davidberard98 replied:
@albanD

> update the doc to explain what this argument is doing

where is this doc? I couldn't find any docstring for this

albanD replied:
Ho wow, I assumed we had a docstring or a cpp comment for this :o 
Well, I guess that's fine, feel free to leave a small comment in cpp above the function saying what this does as a start

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -803,7 +803,11 @@ static PyObject* THPVariable_make_wrapper_subclass(
+    auto override_storage_size = r.toSymIntOptional(14);
+
+    if (override_storage_size.has_value()) {
+      size_bytes = override_storage_size.value();
```
I think we should still do the computation below to ensure that "storage_size >= size_bytes". It would be a completely invalid Tensor if it indexes values outside of its storage.

davidberard98 replied:
my entire goal here is just to skip the computation of size_bytes, because it fails for sizes that include a NestedInt. I don't really have a preference for how we do it - do you have a suggestion?

before this PR, NT uses size_bytes=0 here. so I imagine that the current implementation isn't any worse than before, at least for NT? I think your previous suggestion was also to use `size_bytes=1`

albanD replied:
Why does it fail for NestedInt? You should be able to compute the size of the values storage (assuming it's stored contiguously) right?

davidberard98 replied:
https://github.com/pytorch/pytorch/blob/ebb8905e0c9c21f80a2428f43c9478324d1b4282/aten/src/ATen/EmptyTensor.cpp#L116

Addition is not implemented for NestedInt (and I'm not really sure we can/should implement it for NestedInt)

albanD replied:
I think we could discuss that more but it is beyond this PR for sure!
I would argue that it would be so cool to be able to do something like:
```python
# Assume I want a new Tensor with the first two dimensions collapsed
nested_size = my_nested_tensor.size()
nested_size[1] = nested_size[0] * nested_size[1]
nested_size.pop(0)
my_new_nested_tensor = torch.zeros(nested_size)
```

davidberard98 replied:
interesting idea! filed an issue here: https://github.com/pytorch/pytorch/issues/125940 - perhaps Joel or Jeffrey have an idea of how to do this

albanD commented on torch/csrc/profiler/standalone/execution_trace_observer.cpp:
```diff
@@ -86,7 +86,7 @@ inline std::string getValueShape(
     auto& tensor = val.toTensor();
     if (tensor.defined() &&
-        tensor.unsafeGetTensorImpl()->does_not_have_symbolic_sizes_strides()) {
+        !tensor.unsafeGetTensorImpl()->has_symbolic_sizes_strides()) {
```
What is the difference between the two here? Should we delete the other and flip the flag for everything?

davidberard98 replied:
yes, we should delete the old one. It is only used here, was introduced a few weeks ago

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. It passes in the actual sizes when we call `_make_wrapper_subclass` - this is the change that makes the subclass register as `has_symbolic_sizes_strides() == True`
2. It adds a field to `_make_wrapper_subclass` where an explicit `numel` can be provided. This allows us to skip the numel computation for the storage, which previously fails due to arithmetic on NestedInts.
3. Implements `aten::numel` for NJT - this is separate from the overridden numel in `make_wrapper_subclass` for now. Note also that this means that we leave `dispatch_sizes_strides_policy="sizes"`, so that we call into the custom `numel` implementation (as well as `sizes` and `strides`), because `numel` cannot currently be computed from `sizes` for NJT.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. It passes in the actual sizes when we call `_make_wrapper_subclass` - this is the change that makes the subclass register as `has_symbolic_sizes_strides() == True`
2. It adds a field to `_make_wrapper_subclass` where an explicit `numel` can be provided. This allows us to skip the numel computation for the storage, which previously fails due to arithmetic on NestedInts.
3. Implements `aten::numel` for NJT - this is separate from the overridden numel in `make_wrapper_subclass` for now. Note also that this means that we leave `dispatch_sizes_strides_policy="sizes"`, so that we call into the custom `numel` implementation (as well as `sizes` and `strides`), because `numel` cannot currently be computed from `sizes` for NJT.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57225736).

albanD reviewed (approved):
Thanks for the detailed doc! Sounds good to me!

davidberard98 committed:
Update on "[NT] Make NestedTensor register as having symbolic sizes/strides"


Fixes #123698

This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

1. It passes in the actual sizes when we call `_make_wrapper_subclass` - this is the change that makes the subclass register as `has_symbolic_sizes_strides() == True`
2. It adds a field to `_make_wrapper_subclass` where an explicit `numel` can be provided. This allows us to skip the numel computation for the storage, which previously fails due to arithmetic on NestedInts.
3. Implements `aten::numel` for NJT - this is separate from the overridden numel in `make_wrapper_subclass` for now. Note also that this means that we leave `dispatch_sizes_strides_policy="sizes"`, so that we call into the custom `numel` implementation (as well as `sizes` and `strides`), because `numel` cannot currently be computed from `sizes` for NJT.

Note also that this depends on #121361, because calling TensorImpl::set_sizes_and_strides() tries to clone the sizes into the tensor, which means that we need `clone` to be implemented on NestedInt.

Differential Revision: [D57225736](https://our.internmc.facebook.com/intern/diff/D57225736)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57225736).

davidberard98 commented:
@pytorchbot merge

jbschlosser commented:
> This PR makes TensorImpl::has_symbolic_sizes_strides return false for NestedTensors.

Is this PR desc accurate or should this say `return True`?

davidberard98 commented:
@jbschlosser whoops - yeah, you're correct, it should say `return True`.
